src/client.rs

```diff
@@ -473,42 +472,50 @@ impl Web3 {
             gp
         } else {
             let gas_price = self.eth_gas_price().await?;
-            let f32_gas = gas_price.to_u128();
-            if let Some(v) = f32_gas {
+            if gas_price.sig_bits() <= 128 {
                 // convert to f32, multiply, then convert back, this
                 // will be lossy but you want an exact price you can set it
-                ((v as f32 * gas_price_multiplier) as u128).into()
+                Uint256::from_u128(
+                    (gas_price.resize_to_u128() as f64 * (gas_price_multiplier as f64)) as u128,
+                )
             } else {
                 // gas price is insanely high, best effort rounding
                 // perhaps we should panic here
-                gas_price * (gas_price_multiplier.round() as u128).into()
+                gas_price
+                    .checked_mul(Uint256::from_u128(gas_price_multiplier.round() as u128))
+                    .unwrap()
             }
         };
 
         let mut gas_limit = if let Some(gl) = gas_limit {
             gl
         } else {
-            let gas = self
-                .simulated_gas_price_and_limit(our_balance.clone())
-                .await?;
+            let gas = self.simulated_gas_price_and_limit(our_balance).await?;
             self.eth_estimate_gas(TransactionRequest {
                 from: Some(own_address),
                 to: to_address,
-                nonce: Some(nonce.clone().into()),
+                nonce: Some(nonce.into()),
                 gas_price: Some(gas.price.into()),
                 gas: Some(gas.limit.into()),
-                value: Some(value.clone().into()),
+                value: Some(value.into()),
                 data: Some(data.clone().into()),
             })
             .await?
         };
 
         // multiply limit by gasLimitMultiplier
-        let gas_limit_128 = gas_limit.to_u128();
-        if let Some(v) = gas_limit_128 {
-            gas_limit = ((v as f32 * gas_limit_multiplier) as u128).into()
+        if gas_limit.sig_bits() <= 128 {
+            // convert to f32, multiply, then convert back, this
+            // will be lossy but you want an exact price you can set it
+            gas_limit = Uint256::from_u128(
+                (gas_limit.resize_to_u128() as f64 * (gas_limit_multiplier as f64)) as u128,
+            );
         } else {
-            gas_limit *= (gas_limit_multiplier.round() as u128).into()
+            // gas price is insanely high, best effort rounding
+            // perhaps we should panic here
+            gas_limit = gas_limit
+                .checked_mul(Uint256::from_u128(gas_limit_multiplier.round() as u128))
+                .unwrap()
```

Comment:
same as in clarity, a lot of unwraps here, would this make for a potential attack surface for our orchestrator?

Reply:
There is _less_ attack surface. Unchecked arithmetic wraps in release mode, and if an attacker was able to manipulate one of the two inputs of a multiplication to be large enough, they could get the output to be almost any number they wished. I'm not familiar enough with the code to know the implications, but I would rather that the relayer panic and halt than something worse happening.

Reply:
well that's the thing, we could/should then return an error so that (if needed) we exit out of it cleanly (or try to recover in some way) instead of panicking and breaking everything

Reply:
I have decided to return an error for that branch of the code, since the total supply in the world is much less than what a u128 can handle.